Use entire ringbuffers during split brain healing

mmedenjak opened the pull request:
Previously single items were provided to the merge policy. The same item
was found by comparing the equality of the items in the ringbuffers.
This is not correct in all scenarios. To provide more flexibility and
allow the user to write a more complex merge policy, we provide the
entire ringbuffers to the user during split brain healing.

Part of https://github.com/hazelcast/hazelcast/issues/12803
Part of https://github.com/hazelcast/hazelcast/issues/12804

mmedenjak commented:
run-lab-run

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/MergeOperation.java:
```diff
@@ -25,74 +25,107 @@
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
+        RingbufferContainer<Object, Object> existingContainer = service.getContainerOrNull(getPartitionId(), namespace);
+
+        final RingbufferMergeTypes mergingValue = getMergingType(mergingRingbuffer);
```
This is the only merge operation which contains the merge logic, instead of delegating it to the container. I guess it's due to the `service.destroyContainer(getPartitionId(), namespace);` and the `RingbufferContainer` not having a reference to its service.

mmedenjak replied:
Yes, I'll probably change this to get in line with your current IQueue merge PR

Donnerbart replied:
This comment is outdated by now. Most of the data structures have no reference to the service in their container. Eventually there will be more data structures where the merge logic will be in the operation. So it looks like this is the way to go.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/MergeOperation.java:
```diff
@@ -25,74 +25,107 @@
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
+        final RingbufferMergeData<Object> mergingData = new RingbufferMergeData<Object>(ringbuffer);
+        final RingbufferMergeTypes mergingValue = createMergingValue(serializationService, mergingData);
+        serializationService.getManagedContext().initialize(mergingValue);
```
This should not be needed. We just need to call this on merge policies or merging values, which have been deserialized. The `createMergingValue()` already sets the `SerializationService` for pure local merging values.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/MergeOperation.java:
```diff
@@ -25,74 +25,107 @@
+     * @return the merge type corresponding to the given ringbuffer
+     */
+    private RingbufferMergeTypes getMergingType(Ringbuffer<Object> ringbuffer) {
+        final SerializationService serializationService = getNodeEngine().getSerializationService();
```
You could pass the `SerializationService` as parameter, so you don't have to retrieve it twice.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/spi/impl/merge/RingbufferMergeData.java:
```diff
@@ -0,0 +1,210 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl.merge;
```
The `RingbufferMergeData` is exposed to the user as `MergingValue.getValue()` and can be created by the user as return value of the merge policy. Is that correct? It should not be in an `impl` then.

mmedenjak replied:
I'll move it but I'm uncertain about the naming and the API:
- is `RingbufferMergeData` the right name? Can it just be `Ringbuffer` in the merge package? Can it be `MergingRingbuffer`, etc. 
- can we somehow avoid having the capacity as the argument? Does it need to implement `Iterable<E>` or `Collection<E>`? Are some methods missing or are not needed?

Donnerbart replied:
We already have `Ringbuffer` twice, which is already confusing when picking the right import (maybe the one in `impl` should also be renamed?).

mmedenjak replied:
Sure, any suggestions? I don't have anything good in mind - `InternalRingbuffer`?

tkountis commented on hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/MergeOperation.java:
```diff
@@ -25,74 +25,107 @@
+        final RingbufferMergeData<Object> resultData = mergePolicy.merge(mergingValue, existingValue);
+
+        if (resultData == null) {
+            service.destroyContainer(getPartitionId(), namespace);
```
I see that `destroyDistributedObject` releases other resources too, such as listeners. Don't we need that here? (similar for backup OP)

mmedenjak replied:
I'm just in the process of changing this and getting it in line with the [queue merge implementation](https://github.com/hazelcast/hazelcast/pull/12802/files#diff-05b88fcce532485445fb1aa3157a57abR1128). OTOH I don't like that the `destroyDistributedObject` is called. What if the merge policy wants to keep some partitions but drop some other partitions?

tkountis replied:
:+1: lets revisit once changed

mmedenjak replied:
Hm, this was a brain fart - there is only one ringbuffer on one partition (unless it's an event journal ringbuffer) and if the user returns `null` we can simply destroy the object.

tkountis commented on hazelcast/src/test/java/com/hazelcast/ringbuffer/RingbufferMergeIntegerValuesMergePolicy.java:
```diff
@@ -0,0 +1,46 @@
+        implements SplitBrainMergePolicy<RingbufferMergeData<Object>, RingbufferMergeTypes> {
+
+    @Override
+    public RingbufferMergeData<Object> merge(RingbufferMergeTypes mergingValue, RingbufferMergeTypes existingValue) {
```
Can we make it fail-fast if not Integer data type?

mmedenjak replied:
Actually this merge policy is used to select out integer values from a mixed type (Integer & String) ringbuffers, both on the existing and the merging side. We could make it throw an exception, of course. How would the test look then?

Donnerbart commented on hazelcast/src/test/java/com/hazelcast/ringbuffer/RingbufferSplitBrainTest.java:
```diff
@@ -72,25 +81,47 @@
                 {BINARY, PassThroughMergePolicy.class},
                 {BINARY, PutIfAbsentMergePolicy.class},
 
```
Could you also add a test for the new `RemoveValuesMergePolicy` (its part of `SplitBrainTestSupport`)? This should test the new code path, when the merge policy returns `null`. I found a heck of bugs in my implementations with it.

mmedenjak replied:
I added that policy and I also added a ringbuffer store as you did for the queue which actually uncovered a nasty bug so kudos for including the queue store in the tests :)

mmedenjak commented:
@tkountis @Donnerbart redid the implementation. Here are the changes:
- made it more in line with the IQueue merging (https://github.com/hazelcast/hazelcast/pull/12802)
- used `destroyDistributedObject` instead of `destroyContainer` when destroying container
- improved test in according to the queue split brain test (added store, more policies)
- fixed a bug including the ringbuffer store and migrations

The differences between the queue and ringbuffer implementation:
- in ringbuffer, sequences are more important. We give them to the user and we use the returned sequences to set the resulting ringbuffer
- we also set the items in the ringbuffer store on merge to keep the store and in-memory data in-sync
- the queue tests test the queue when the in-memory format is OBJECT. For ringbuffer, we test both OBJECT and BINARY which complicates the test a bit since the data in the ringbuffer store is now a byte array
- the ringbuffer merge operation avoids creating a container if there isn't a container already and the merge result is `null`. As a simplification, we can always get-or-create a container and immediately destroy it if the merge result is `null`. For now, I've added this optimisation but I can remove it if you think it complicates the operation too much.

Please take an another look.

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferContainer.java:
```diff
@@ -141,8 +147,10 @@ private void initRingbufferStore(ClassLoader configClassLoader) {
                 final long storeSequence = store.getLargestSequence();
-                ringbuffer.setTailSequence(storeSequence);
-                ringbuffer.setHeadSequence(storeSequence + 1);
+                if (tailSequence() < storeSequence) {
```
This was the bug with the ringbuffer store and the migrations. It goes like this - the ringbuffer would be migrated with some head and tail sequence (e.g. 0 and 100). After this, it would invoke the store and ask what's the highest sequence. The store returns 100 (since it stored all items) and then the ringbuffer unconditionally set the head and tail sequence to 100 and 101, thus emptying the in-memory state. 
This does not affect correctness but we unnecessarily lost in-memory items. The split brain tests cover this issue as the issue was detected because of the use of the ringbuffer store in the split brain tests, thanks to @Donnerbart

mmedenjak commented:
run-lab-run
